### PR TITLE
SI-3346 implicit parameters can now guide implicit view inference

### PR DESCRIPTION
**Pull request by xeno-by**
This simple patch makes it possible for implicit views to benefit from
fundep-guided type inference, eliminating a nasty special case in
implicit inference.

Here are the changes that I had to apply to the tests (they exposed
quite a number of interesting questions that I was happy to answer):

1) neg/t5845.scala now works, so I moved it to pos. That easily makes sense,
because the test was just a canary to track previous state of things.

2) neg/divergent_implicit.scala and neg/t5578.scala changed their messages
to less informative ones, because inapplicable implicit views now get
disqualified early and therefore don't display their error messages
to the user. This is an unfortunate but necessary byproduct of this
change, and one can argue that the behavior is now completely consistent
with implicit vals (that also don't explain why this or that implicit val
got disqualified, but rather display a generic implicit value not found
message).

3) scaladoc/implicits-chaining.scala and scaladoc/implicits-base.scala.
Immediate culling of apriori inapplicable implicit views messes things up
for Scaladoc, because it's interested in potentially applicable views,
having infrastructure to track various constraints (type bounds, requirements
for implicit parameters, etc) that guide applicability of views and then
present it to the user. Therefore, when scaladoc is detected, implicit search
reverts to the old behavior.

4) We still cannot have Jason's workaround to type constructor inference
mentioned in comments to SI-3346, because it's not really about implicit 
parameters of implicit views, but about type inference flowing from the 
implicit parameter list to a preceding parameter list in order to affect 
inference of an implicit view. This is something that's still too ambitious.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @adriaanm @retronym 

/cc @paulp @hubertp @blaisorblade @milessabin @gkossakowski @harrah @szeiger

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -603,7 +603,23 @@ trait Implicits {
...
-            )
+            ) match {
+              // try to infer implicit parameters immediately in order to:
+              //   1) guide type inference for implicit views
+              //   2) discard ineligible views right away instead of risking spurious ambiguous implicits
+              //
+              // this is an improvement of the state of the art that brings consistency to implicit resolution rules
+              // (and also helps fundep materialization to be applicable to implicit views)
+              //
+              // there's one caveat though. we need to turn this behavior off for scalaodc
```
Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: scalaodc->scaladoc?

Comment (Blaisorblade) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: (I can't add anything more relevant on the code itself, yet, so sorry if I only provide nitpicking).

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Thanks! Fixed.

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 2]
```diff
@@ -603,7 +603,23 @@ trait Implicits {
...
+              // this is an improvement of the state of the art that brings consistency to implicit resolution rules
+              // (and also helps fundep materialization to be applicable to implicit views)
+              //
+              // there's one caveat though. we need to turn this behavior off for scaladoc
+              // because scaladoc usually doesn't know the entire story
+              // and is just interested in views that are potentially applicable
+              // for instance, if we have `class C[T]` and `implicit def conv[T: Numeric](c: C[T]) = ???`
+              // then Scaladoc will give us something of type `C[T]`, and it would like to know
+              // that `conv` is potentially available under such and such conditions
+              case tree if isImplicitMethodType(tree.tpe) && !isScalaDoc => applyImplicitArgs(tree)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:620]: isn't `isScalaDoc` ruled out by the `if (isLocal && !isScalaDoc) {` above?

**Comment by adriaanm** [comment]
This worries me. We should at the very least rebuild scalaz, specs2 and scalatest to make sure this doesn't break them.

**Comment by adriaanm** [comment]
(I would love to remove this discrepancy, though!)

**Comment by xeno-by** [comment]
@adriaanm How is this done these days? Should I learn dbuild or a manual rebuild would be enough?

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala (hunk 2 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:620]: That's a funny artifact of how git presented this particular diff. Even though this pattern match appears nested into that condition, it is not :)

**Comment by adriaanm** [comment]
We'll soon have a dbuild job as part of PR validation, but I'm completely
swamped the next two weeks.
We can probably get one set up earlier akin to scala-checkin-manual.

Realistically, this is a big change and I don't think we can still take
this on for M5. Let's see about M6.

**Comment by larsrh** [comment]
Also, scalaz 7.1.x doesn't currently build against M4 (could probably be fixed on our side, though, but I didn't get around to it yet).

**Comment by xeno-by** [comment]
Oops, sorry my bad. I wanted to say that I agree about M6 and misclicked the button.

**Comment by hubertp** [comment]
@larsrh Is that still the bug you showed me at the conference but I didn't have time to have a look at before going on vacation?

**Comment by larsrh** [comment]
@hubertp Yes, it's that bug. Can be fixed by adding explicit type parameters, though. I failed to produce a minimal example, though (apparently doesn't happen in isolation).

**Comment by szeiger** [comment]
This looks awesome (conceptually - I don't really understand the change to the compiler) and could lead to some big simplifications in Slick's Lifted Embedding.

**Comment by hubertp** [comment]
@larsrh yeah, I know. I really tried to minimize this thins and created very similar structure but still failed to reproduce. Really confusing.

**Comment by larsrh** [comment]
For the records, scalaz now compiles with 2.11.0-M4: scalaz/scalaz@c0f3a29fb9 (albeit only without tests; try `sbt example/compile`).

**Comment by retronym** [comment]
> albeit only without tests

Because of a failure, or just because of unavailability of specs2?

**Comment by larsrh** [comment]
@retronym Latter. I backported the relevant commit to the 7.0.x branch. @etorreborre, anything else needed to get a specs2 1.12.x against 2.11.x?

**Comment by etorreborre** [comment]
specs2 1.12.x is only compiled against 2.9.x. 1.13 starts being available for 2.10.x. Do you need 1.13 for 2.11.x then?

**Comment by gkossakowski** [comment]
@xeno-by What's the status of this PR?

**Comment by xeno-by** [comment]
Waiting for: 1) a review, 2) a decision on which libraries and in what manner we should test it.

**Comment by szeiger** [comment]
I was hoping this would take care of http://pastie.org/8284802 but it does not

**Comment by szeiger** [comment]
For the record, Slick (master) compiles successfully against this patch.

**Comment by Blaisorblade** [comment]
> I was hoping this would take care of http://pastie.org/8284802 but it does not

I was checking the expected behavior. Since I can't comment there, here's what implicit resolution should produce (from an interaction with the 2.10.2 REPL):

``` scala
scala> def bar[F, T, G](f: F)(implicit shape: Shape[Flat, F, T, G]) = shape
bar: [F, T, G](f: F)(implicit shape: Shape[Flat,F,T,G])Shape[Flat,F,T,G]

scala> bar(null: ((Column[Int], Column[Int]), Suppliers))(tuple2Shape(tuple2Shape, columnBaseShape))
res13: Shape[Flat,((Column[Int], Column[Int]), Suppliers),((Int, Int), (Int, Int)),((Column[Int], Column[Int]), Suppliers)] = null
```

**Comment by gkossakowski** [comment]
@retronym, @adriaanm: ping on pending review

**Comment by retronym** [comment]
I'm still scared by this one (bug fixes in implcit resolution can silently change the meaning of existing code.) I must first defer to @adriaanm to review this on its own merits, and if he's happy with that, we can weigh up the potential for breakage vs the improvements in power + regularity that this brings.

**Comment by Blaisorblade** [comment]
Meanwhile, specs2 and scalaz still need porting to HEAD, to be tested against this commit.

@larsrh, @etorreborre: ping. What's the status of this porting effort?

To summarize the current situation, last message was:

> specs2 1.12.x is only compiled against 2.9.x. 1.13 starts being available for 2.10.x. Do you need 1.13 for 2.11.x then?

In fact, specs2 1.12.3 _is_ [available for 2.10](http://mvnrepository.com/artifact/org.specs2/specs2_2.10/1.12.3), unlike earlier said (this explains how scalaz _is_ getting tested currently), and that's what Scalaz [uses](https://github.com/scalaz/scalaz/blob/scalaz-seven/project/build.scala#L291).

Any chance to disentangle this?

**Comment by Blaisorblade** [comment]
@larsh: on the light side, there's only one documented breaking change in all releases between specs2 1.12 and 2.2:
http://notes.implicit.ly/tagged/specs2

**Comment by etorreborre** [comment]
@Blaisorblade 

> In fact, specs2 1.12.3 is available for 2.10, unlike earlier said (this explains how scalaz is getting tested currently), and that's what Scalaz uses.

Right, I forgot about that. 

> Any chance to disentangle this?

If Scalaz can be pushed without tests (i.e. without needing a specs2 dependency) then I can push a specs2 version against and then we can close the loop by using the new specs2 version to finish compiling / testing Scalaz fully.

**Comment by gkossakowski** [comment]
This can't merge anymore. I'm closing it. Please reopen a new PR with rebased commits and refer to this PR so both are linked.

**Comment by xeno-by** [comment]
I would like to get a review first please, so that we don't have to rebase again it if a timeout happens again. Do you think that would be possible? If yes, please reopen this PR.

**Comment by gkossakowski** [comment]
@xeno-by: good point. I'll reopen and assign to @adriaanm.

**Comment by retronym** [comment]
We can still review this on a closed PR. I'm trying to shrink the PR queue to mitigate the build kitteh craziness. @adriaanm don't take the closure as reason not to review :)

**Comment by xeno-by** [comment]
Guys, since M6 is very close, could we extend the deadline for this pull request to M7? It would be a shame to hastily reject it just because M6 is supposed to be the last development milestone, and we don't have time to have a discussion before M6.

**Comment by Blaisorblade** [comment]
:+1:

**Comment by adriaanm** [comment]
Yes, M7 is okay with me (M8 is not)

**Comment by xeno-by** [comment]
As per today's core meeting, the code looks good, and we just need to make sure that we don't get regression in community libraries, such as scalaz, spray, specs2, shapeless.

**Comment by xeno-by** [comment]
Superseded by https://github.com/scala/scala/pull/3030

**Pull request closed without merging**